src/main/scala/org/hammerlab/guacamole/alignment/AffineGapPenaltyAlignment.scala

```diff
@@ -0,0 +1,159 @@
+   *
+   * @param alignmentStates (sequenceLength + 1) x (referenceLength + 1) matrix of best transition at each step
+   * @param alignmentScores (sequenceLength + 1) x (referenceLength + 1) matrix of alignment score at each step
+   * @param sequenceLength Length of the input sequence to align
+   * @param referenceLength Length of reference sequence to align against
+   * @return An alignment path and score
+   */
+  def findBestPath(alignmentStates: DenseMatrix[AlignmentState],
+                   alignmentScores: DenseMatrix[Double],
+                   sequenceLength: Int,
+                   referenceLength: Int): ReadAlignment = {
+
+    @tailrec
+    def recFindBestPath(sequenceIdx: Int,
+                        referenceIdx: Int,
+                        currentPath: List[AlignmentState]): List[AlignmentState] = {
+
+      val lastMove = alignmentStates(sequenceIdx, referenceIdx)
+      if (sequenceIdx == 1 && referenceIdx == 1) {
+        lastMove :: currentPath
+      } else {
+        lastMove match {
+          case AlignmentState.Match | AlignmentState.Mismatch =>
+            recFindBestPath(sequenceIdx - 1, referenceIdx - 1, lastMove :: currentPath)
+          case AlignmentState.Insertion =>
+            recFindBestPath(sequenceIdx - 1, referenceIdx, lastMove :: currentPath)
+          case AlignmentState.Deletion =>
+            recFindBestPath(sequenceIdx, referenceIdx - 1, lastMove :: currentPath)
+        }
+      }
+    }
+
+    val alignments = recFindBestPath(sequenceLength, referenceLength, List.empty)
+    ReadAlignment(alignments, argmin(alignmentScores.t(::, sequenceLength)))
+  }
+
+  /**
+   * Produces an alignment of an input sequence against a reference sequence
+   *
+   * @param sequence Input sequence to align
+   * @param reference Reference sequence to align against
+   * @param mismatchProbability Penalty to having a mismatch in the alignment
+   * @param openGapProbability Penalty to start an insertion or deletion in the alignment
+   * @param closeGapProbability Penalty to end in an insertion or deletion in the alignment
+   * @return An alignment path and score
+   */
+  def align(sequence: Seq[Byte],
+            reference: Seq[Byte],
+            mismatchProbability: Double = math.exp(-4),
+            openGapProbability: Double = math.exp(-6),
+            closeGapProbability: Double = 1 - math.exp(-1)): ReadAlignment = {
+    // TODO: What are the best defaults?
+    // BWA defaults:
+    // -log mismatchProbability = 4
+    // -log openGapProbability = 6
+    // -log (1 - closeGapProbability) = 1
+
+    val sequenceLength = sequence.length
+    val referenceLength = reference.length
+    val (alignmentStates, alignmentScores) = scoreAlignmentPaths(sequence,
+      reference,
+      mismatchProbability,
+      openGapProbability,
+      closeGapProbability)
+
+    findBestPath(alignmentStates, alignmentScores, sequenceLength, referenceLength)
```

Comment:
one thing that just occurred to me: it's interesting that you're building this matrix and then traversing back through it to back out the best path; it could be built in such a way that each element stores the best path to it, saving this second traversal, right?

I assume that would be more memory-consumptive and therefore not worth it, just wanted to confirm that we're on the same page about that.

Also, very much as an aside: it seems like there's a pruned search that could be done here, as a possible optimization, using a "frontier" pqueue that keeps all next-to-be-expanded matrix positions in order of increasing score (starting with just `(0,0) → 0`… I'm sure there is literature about this, I will think about it some more.